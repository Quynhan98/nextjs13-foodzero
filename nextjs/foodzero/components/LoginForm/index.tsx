import { ChangeEvent, FormEvent, memo } from 'react'
import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  Button,
} from '@chakra-ui/react'

// Themes
import { rufina } from '@themes/index'

interface LoginFormProps {
  onSubmitForm?: (e: FormEvent<HTMLFormElement>) => void
  onChangeForm?: (e: ChangeEvent<HTMLInputElement>) => void
  emailError?: string
  passwordError?: string
  isDisable: boolean
}

const LoginForm = ({
  onSubmitForm,
  onChangeForm,
  emailError,
  passwordError,
  isDisable,
}: LoginFormProps) => {
  return (
    <form onSubmit={onSubmitForm}>
      <FormLabel
        fontSize={{ base: 'md', md: 'lg' }}
        fontWeight="bold"
        fontFamily={rufina.style.fontFamily}
        textAlign="center"
        pb={{ base: '58px', '2xl': '115px' }}
      >
        Login
      </FormLabel>
      <FormControl
        boxShadow="2xl"
        display="flex"
        flexDirection="column"
        justifyContent="center"
        isInvalid={!!(emailError || passwordError)}
        padding={{ base: '25px', '2xl': '50px' }}
      >
        <Input
          data-testid="email"
          type="email"
          name="email"
          placeholder="Email"
          onChange={onChangeForm}
        />
        {emailError && (
          <FormErrorMessage fontSize={{ base: 'xxs', '2xl': 'base' }}>
            {emailError}
          </FormErrorMessage>
        )}
        <Input
          data-testid="password"
          type="password"
          name="password"
          placeholder="Password"
          onChange={onChangeForm}
          mt={{ base: '25px', '2xl': '50px' }}
        />
        {passwordError && (
          <FormErrorMessage fontSize={{ base: 'xxs', '2xl': 'base' }}>
            {passwordError}
          </FormErrorMessage>
        )}
        <Button
          mt={{ base: '40px', '2xl': '80px' }}
          type="submit"
          isDisabled={isDisable}
          _hover={{ backgroundColor: 'none' }}
          size="large"
        >
          Login
        </Button>
      </FormControl>
    </form>
  )
}

export default memo(LoginForm)
