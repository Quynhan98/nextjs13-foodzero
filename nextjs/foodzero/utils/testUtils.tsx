import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'

// Themes
import { customTheme } from '@themes/index'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { BookingProvider } from '@contexts/BookingProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ChakraProvider theme={customTheme}>
      <AuthProvider>
        <BookingProvider>{component}</BookingProvider>
      </AuthProvider>
    </ChakraProvider>,
  )
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { customRender as render }
