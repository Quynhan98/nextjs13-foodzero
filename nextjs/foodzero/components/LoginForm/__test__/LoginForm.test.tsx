import React from 'react'

// Utils
import { render, screen, cleanup, fireEvent } from '@utils/testUtils'

// Components
import LoginForm from '@components/LoginForm'
import {
  INVALID_EMAIL_FORMAT,
  INVALID_PASSWORD_FORMAT,
} from '@constants/errorMessage'

describe('LoginForm', () => {
  const accountMock = {
    email: 'foodzero@gmail.com',
    password: '123456789',
  }

  const props = {
    onChangeForm: jest.fn(),
    onSubmitForm: jest.fn(),
    error: '',
    isDisable: false,
  }

  beforeEach(() => {
    render(<LoginForm {...props} />)
  })

  afterEach(() => {
    cleanup()
  })

  it('LoginForm matches snapshot', () => {
    const { container } = render(<LoginForm {...props} />)

    expect(container).toMatchSnapshot()
  })

  it('Should call onChange input email', () => {
    const email = screen.getByTestId('email')

    fireEvent.change(email, {
      target: { value: accountMock.email },
    })

    expect(props.onChangeForm).toBeCalled()
    expect(email).toHaveValue(accountMock.email)
  })

  it('Should call onChange input password', () => {
    const password = screen.getByTestId('password')

    fireEvent.change(password, {
      target: { value: accountMock.password },
    })

    expect(props.onChangeForm).toBeCalled()
    expect(password).toHaveValue(accountMock.password)
  })

  it('Should call submit form', () => {
    const button = screen.getByRole('button')

    fireEvent.submit(button)

    expect(props.onSubmitForm).toBeCalled()
  })

  it('Should render ErrorMessage', () => {
    render(
      <LoginForm
        {...props}
        emailError={INVALID_EMAIL_FORMAT}
        passwordError={INVALID_PASSWORD_FORMAT}
      />,
    )

    expect(screen.getByText(INVALID_EMAIL_FORMAT)).toHaveTextContent(
      INVALID_EMAIL_FORMAT,
    )

    expect(screen.getByText(INVALID_PASSWORD_FORMAT)).toHaveTextContent(
      INVALID_PASSWORD_FORMAT,
    )
  })
})
