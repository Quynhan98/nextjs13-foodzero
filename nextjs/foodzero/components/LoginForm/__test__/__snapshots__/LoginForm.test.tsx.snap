// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm LoginForm matches snapshot 1`] = `
<div>
  <form>
    <label
      class="chakra-form__label css-78n0uo"
    >
      Login
    </label>
    <div
      class="chakra-form-control css-xj5wq"
      role="group"
    >
      <input
        class="chakra-input css-1njo91c"
        data-testid="email"
        id="field-:r1:"
        name="email"
        placeholder="Email"
        type="email"
      />
      <input
        class="chakra-input css-1abompr"
        data-testid="password"
        id="field-:r1:"
        name="password"
        placeholder="Password"
        type="password"
      />
      <button
        class="chakra-button css-157j7mr"
        type="submit"
      >
        Login
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
